### PR TITLE
fix(index.d.ts): improve typings for virtuals

### PR DESCRIPTION
**Pull request by thiagokisaki**
**Summary**
Created `VirtualTypeOptions` interface, improved `VirtualType` class and adjusted `Schema` methods related to virtuals.

**Examples**
I added a test file in `test/typescript/virtuals.ts`.

### CODE REVIEW

**Comment by thiagokisaki** [comment]
These types are based on the following pages from mongoose docs:
* https://mongoosejs.com/docs/guide.html#virtuals
* https://mongoosejs.com/docs/tutorials/virtuals.html
* https://mongoosejs.com/docs/populate.html#populate-virtuals
* https://mongoosejs.com/docs/api/virtualtype.html

**Review by vkarpov15 — APPROVED** [review]
Thanks :+1:

File: index.d.ts [hunk 1]
```diff
@@ -1618,18 +1618,68 @@ declare module 'mongoose' {
...
+    applyGetters(value: any, doc: D): any;
+
     /** Applies setters to `value`. */
-    applySetters(value: any, doc: Document): any;
+    applySetters(value: any, doc: D): any;
 
     /** Adds a custom getter to this virtual. */
-    // eslint-disable-next-line @typescript-eslint/ban-types
-    get(fn: Function): this;
+    get(fn: (this: D, value: any, virtual: this, doc: D) => any): this;
```
Comment (vkarpov15) [inline index.d.ts:1679]: Unfortunately I had to undo these changes. For some reason adding this generic param breaks defining `new Schema()` without a doc type and then using that schema for [`discriminator()`](https://mongoosejs.com/docs/api/model.html#model_Model.discriminator).
